# **Research Question**


**Analyze the relationship between the number and type of educational facilities, the proportion of the population of the elderly and disabled, and the unemployment rate, especially for minors, young people, and the elderly. The machine learning method is used to predict the community unemployment rate and determine which communities have a significant relationship with educational facilities and demographic characteristics.**

分析教育设施数量和类型、老年人和残障人士的人口比例与失业率之间的关系，特别是针对未成年人、年轻人和老年人群体。并通过机器学习方法预测社区失业率，确定哪些社区的失业率与教育设施和人口特征有明显关系。

In [18]:
import pandas as pd

# **Data Preprocessing**

## **1. Extract the required columns:**


**Demographic data:** elderly, minors, young people, people with disabilities in 2007 and 2012
* ERP age 65-69, persons, ERP age 70-74, persons, ERP age 75-79, persons, ERP age 80-84, persons, ERP age 85+, persons
* ERP age 0-4, persons, ERP age 5-9, persons, ERP age 10-14, persons, ERP age 15-19, persons, ERP age 20-24, persons
* Unpaid carer to person with disability, persons

**Educational facilities:**
* Primary Schools, Secondary Schools, TAFE students, University students

**Unemployment rate:**
* Unemployed, %

In [19]:
# Read the file
file_path = 'communities.csv'
data = pd.read_csv(file_path)

In [36]:
# Define the columns to extract for 2012
columns_2012 = [
    '2012 ERP age 65-69, persons', '2012 ERP age 70-74, persons', '2012 ERP age 75-79, persons',
    '2012 ERP age 80-84, persons', '2012 ERP age 85+, persons', '2012 ERP age 0-4, persons',
    '2012 ERP age 5-9, persons', '2012 ERP age 10-14, persons', '2012 ERP age 15-19, persons',
    '2012 ERP age 20-24, persons'
]

# Define the columns to extract for 2007
columns_2007 = [
    '2007 ERP age 65-69, persons', '2007 ERP age 70-74, persons', '2007 ERP age 75-79, persons',
    '2007 ERP age 80-84, persons', '2007 ERP age 85+, persons', '2007 ERP age 0-4, persons',
    '2007 ERP age 5-9, persons', '2007 ERP age 10-14, persons', '2007 ERP age 15-19, persons',
    '2007 ERP age 20-24, persons'
]

# Define the columns with no specific year mentioned
columns_no_year = [
    'Unpaid carer to person with disability, persons', 'Primary Schools', 'Secondary Schools',
    'TAFE students', 'University students', 'Unemployed, %'
]

# Extract the data for 2012, 2007, and columns with no year
selected_data_2012 = data[columns_2012]
selected_data_2007 = data[columns_2007]
selected_data_no_year = data[columns_no_year]

# Concatenate the 2007, 2012, and no-year data into a single DataFrame for comparison
selected_data = pd.concat([selected_data_2007, selected_data_2012, selected_data_no_year], axis=1)

# Rename the columns to clearly differentiate between 2007 and 2012 data
selected_data.columns = [
    '2007 age 65-69', '2007 age 70-74', '2007 age 75-79', '2007 age 80-84', '2007 age 85+',
    '2007 age 0-4', '2007 age 5-9', '2007 age 10-14', '2007 age 15-19', '2007 age 20-24',
    '2012 age 65-69', '2012 age 70-74', '2012 age 75-79', '2012 age 80-84', '2012 age 85+',
    '2012 age 0-4', '2012 age 5-9', '2012 age 10-14', '2012 age 15-19', '2012 age 20-24',
    'Unpaid carer to person with disability', 'Primary Schools', 'Secondary Schools', 
    'TAFE students', 'University students', 'Unemployed % 2012'
]

# Display the first few rows of the dataset to check the format
selected_data.head()

2007 age 65-69  2007 age 70-74  2007 age 75-79  2007 age 80-84  \
0      108.000000       98.000000       87.000000       77.000000   
1      119.444444       91.777777      115.666665      117.111110   
2             NaN             NaN             NaN             NaN   
3       48.936521       48.098261       41.030434       19.546957   
4       20.066666       19.199999       12.666666       13.733333   

   2007 age 85+  2007 age 0-4  2007 age 5-9  2007 age 10-14  2007 age 15-19  \
0     73.000000    208.000000    114.000000      152.000000      156.000000   
1    118.888886    194.999996    283.444439      300.111106      272.111109   
2           NaN           NaN           NaN             NaN             NaN   
3     24.721739     69.990435     69.550434       58.521739       49.214782   
4      6.866667     66.533328     40.266663       36.799997       44.799997   

   2007 age 20-24  ...  2012 age 5-9  2012 age 10-14  2012 age 15-19  \
0      397.000000  ...           161             123             181   
1      262.999998  ...           273             303             311   
2             NaN  ...            15              34              22   
3       40.002609  ...            77              67              57   
4       51.799996  ...            35              32              26   

   2012 age 20-24  Unpaid carer to person with disability  Primary Schools  \
0             462                                     393                1   
1             279                                     404                3   
2              11                                      52                0   
3              26                                     120                0   
4              29                                      63                0   

   Secondary Schools  TAFE students  University students  Unemployed % 2012  
0                  0             95                  431           4.677364  
1                  1             56                  217           3.811434  
2                  0              5                   <5           5.921053  
3                  0             22                   19           6.734694  
4                  0              8                   20           2.922078  

[5 rows x 26 columns]

## **2. Handle Missing Values**

In [37]:
# Check the missing value
missing_values = selected_data.isnull().sum()
print("Missing Values:")
print(missing_values)



Missing Values:
2007 age 65-69                            372
2007 age 70-74                            372
2007 age 75-79                            372
2007 age 80-84                            372
2007 age 85+                              372
2007 age 0-4                              372
2007 age 5-9                              372
2007 age 10-14                            372
2007 age 15-19                            372
2007 age 20-24                            372
2012 age 65-69                              0
2012 age 70-74                              0
2012 age 75-79                              0
2012 age 80-84                              0
2012 age 85+                                0
2012 age 0-4                                0
2012 age 5-9                                0
2012 age 10-14                              0
2012 age 15-19                              0
2012 age 20-24                              0
Unpaid carer to person with disability      0
Primary Schools   

1.	Missing Data in 2007:
* The columns 2007 age 65-69 to 2007 age 20-24 each have 372 missing values. These columns represent the population data for different age groups in 2007.
* This suggests that many communities or regions did not report their population statistics for 2007.

2.	Data in 2012:
* The columns 2012 age 65-69 to 2012 age 20-24 have no missing values. These columns are complete and provide population data for various age groups in 2012.
* Since the 2012 data is complete, we can use it to fill in the missing values in the 2007 data.
	
3.	Other Columns with Missing Data:
* The column Unemployed % 2012 has 64 missing values, indicating that some areas did not report their unemployment rate.
* The columns Unpaid carer to person with disability, Primary Schools, Secondary Schools, TAFE students, and University students have no missing values.

In [40]:
# Fill in the missing values for 2007 using the 2012 mean

columns_2007 = [
    '2007 age 65-69', '2007 age 70-74', '2007 age 75-79',
    '2007 age 80-84', '2007 age 85+', '2007 age 0-4', 
    '2007 age 5-9', '2007 age 10-14', '2007 age 15-19', 
    '2007 age 20-24'
]

columns_2012 = [
    '2012 age 65-69', '2012 age 70-74', '2012 age 75-79',
    '2012 age 80-84', '2012 age 85+', '2012 age 0-4', 
    '2012 age 5-9', '2012 age 10-14', '2012 age 15-19', 
    '2012 age 20-24'
]

for col_2007, col_2012 in zip(columns_2007, columns_2012):
    selected_data[col_2007] = selected_data[col_2007].fillna(selected_data[col_2012].mean())

In [41]:
# Fill in the missing values using the mean of the unemployment rate column

selected_data['Unemployed % 2012'] = selected_data['Unemployed % 2012'].fillna(selected_data['Unemployed % 2012'].mean())

## **3. Handle Non-numeric Columns**

In [42]:
# Check the non-numeric columns
non_numeric_columns = selected_data.select_dtypes(exclude=['number']).columns
print("Non-numeric Columns:")
print(non_numeric_columns)

Non-numeric Columns:
Index(['Unpaid carer to person with disability', 'TAFE students',
       'University students'],
      dtype='object')


The non-numeric columns identified in your dataset are:

* Unpaid carer to person with disability
* TAFE students
* University students

These columns should be of numeric type, but are currently treated as non-numeric columns, possibly because of the presence of non-numeric characters (such as commas or other characters) in the data.

In [46]:
# Convert non-numeric columns to numeric, coercing errors to NaN
for col in non_numeric_columns:
    selected_data[col] = pd.to_numeric(selected_data[col], errors='coerce')

# Check if the conversion was successful
selected_data[non_numeric_columns].head()

Unpaid carer to person with disability  TAFE students  University students
0                                   393.0           95.0                431.0
1                                   404.0           56.0                217.0
2                                    52.0            5.0                  NaN
3                                   120.0           22.0                 19.0
4                                    63.0            8.0                 20.0

## **4. Recheck the data set for missing values**

In [47]:
missing_values = selected_data.isnull().sum()
print("Missing Values After Conversion:")
print(missing_values)

Missing Values After Conversion:
2007 age 65-69                              0
2007 age 70-74                              0
2007 age 75-79                              0
2007 age 80-84                              0
2007 age 85+                                0
2007 age 0-4                                0
2007 age 5-9                                0
2007 age 10-14                              0
2007 age 15-19                              0
2007 age 20-24                              0
2012 age 65-69                              0
2012 age 70-74                              0
2012 age 75-79                              0
2012 age 80-84                              0
2012 age 85+                                0
2012 age 0-4                                0
2012 age 5-9                                0
2012 age 10-14                              0
2012 age 15-19                              0
2012 age 20-24                              0
Unpaid carer to person with disability      3
P

**Handle missing values in the column Unpaid carer to person with disability**

Fill with the mean: Replaces the missing value of a column with the mean of that column.

In [49]:
selected_data['Unpaid carer to person with disability'] = selected_data['Unpaid carer to person with disability'].fillna(selected_data['Unpaid carer to person with disability'].mean())

**Handles missing values in TAFE students and University students columns**

The high number of missing values in these two columns, 83 and 105 respectively, could mean that there are no TAFE students or university students in some communities, or that the data is not being reported. I use 0 padding here: Since some areas may indeed have no TAFE or university students, use 0 to represent these areas.

In [50]:
selected_data['TAFE students'] = selected_data['TAFE students'].fillna(0)
selected_data['University students'] = selected_data['University students'].fillna(0)

## **5. Data Scaling**